### PR TITLE
Add `format` for trace ID propagation.

### PR DESCRIPTION
**Pull request by BrynCooke**
When propagating trace ID to subgraph via header is is now possible to specify the format.

```yaml
telemetry:
  exporters:
    tracing: 
      propagation: 
        request: 
          header_name: "my_header"
          format: uuid
```

Note that incoming requests must some form of UUID either with or without dashes.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: .changesets/feat_propagation_format.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+### Add `format` for trace ID propagation. ([PR #5803](https://github.com/apollographql/router/pull/5803))
+
+When propagating trace ID to subgraph via header is is now possible to specify the format.
```
Comment (shorgi) [inline .changesets/feat_propagation_format.md:3]:
```suggestion
The router now supports specifying the format of trace IDs that are propagated to subgraphs via headers.

You can configure the format with the `format` option:
```

File: .changesets/feat_propagation_format.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+telemetry:
+  exporters:
+    tracing: 
+      propagation: 
+        request: 
+          header_name: "my_header"
+          format: uuid
+```
+
+Note that incoming requests must some form of UUID either with or without dashes.
```
Comment (shorgi) [inline .changesets/feat_propagation_format.md:15]:
```suggestion
Note that incoming requests must be some form of UUID, either with or without dashes.

To learn about supported formats, go to [`request` configuration reference](https://apollographql.com/docs/router/configuration/telemetry/exporters/tracing/overview#request-configuration-reference) docs. 
```

File: .changesets/feat_propagation_format.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+When propagating trace ID to subgraph via header is is now possible to specify the format.
+
+```yaml
+telemetry:
+  exporters:
+    tracing: 
+      propagation: 
+        request: 
+          header_name: "my_header"
+          format: uuid
```
Comment (shorgi) [inline .changesets/feat_propagation_format.md:12]:
```suggestion
          # Must be in UUID form, with or without dashes
          format: uuid
```

File: docs/source/configuration/telemetry/exporters/tracing/overview.mdx [hunk 4]
```diff
@@ -144,9 +144,28 @@ telemetry:
...
            header_name: my-trace-id
+           # The format of the trace when propagating to subgraphs.
+           format: uuid
 ```
 
+#### `request` configuration reference
+
+| Option        | Values                                                        | Default                           | Description                         |
+|---------------|---------------------------------------------------------------|-----------------------------------|-------------------------------------|
+| `header_name` |                                                               | Enable or disable stdout logging. |                                     |
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/tracing/overview.mdx:157]: Is this a boolean? Should its value be `boolean` or `true|false`?

File: docs/source/configuration/telemetry/exporters/tracing/overview.mdx [hunk 5]
```diff
@@ -144,9 +144,28 @@ telemetry:
...
+| `format`      | `hexadecimal`\|`open_telemetry`\|`decimal`\|`datadog`\|`uuid` | `hexadecimal`                     | The output format of the `trace_id` |
+
+The formats are:
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/tracing/overview.mdx:160]:
```suggestion
Valid values for `format`:
```

File: docs/source/configuration/telemetry/exporters/tracing/overview.mdx [hunk 6]
```diff
@@ -144,9 +144,28 @@ telemetry:
...
+* `hexadecimal` - 32-character hexadecimal string (e.g. `0123456789abcdef0123456789abcdef`)
+* `open_telemetry` - 32-character hexadecimal string (e.g. `0123456789abcdef0123456789abcdef`)
+* `decimal` - 16-character decimal string (e.g. `1234567890123456`)
+* `datadog` - 16-character decimal string (e.g. `1234567890123456`)
+* `uuid` - 36-character UUID string (e.g. `01234567-89ab-cdef-0123-456789abcdef`)
+
+Note that incoming trace IDs must be in `open_telemetry` or `uuid` format.
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/tracing/overview.mdx:167]:
```suggestion
<Note>

Incoming trace IDs must be in `open_telemetry` or `uuid` format.

</Note>
```

File: docs/source/configuration/telemetry/exporters/tracing/overview.mdx (hunk 5 above)
Comment (BrynCooke) [reply docs/source/configuration/telemetry/exporters/tracing/overview.mdx:160]: Oops, the description is wrong.

File: docs/source/configuration/telemetry/exporters/tracing/overview.mdx (hunk 4 above)
Comment (BrynCooke) [reply docs/source/configuration/telemetry/exporters/tracing/overview.mdx:157]: Oops the description is wrong.

Comment (BrynCooke) [inline docs/source/configuration/telemetry/exporters/tracing/overview.mdx:157]:
```suggestion
| `header_name` |                                                               | The name of the http header to use for propagation. |                                     |
```

Comment (BrynCooke) [inline docs/source/configuration/telemetry/exporters/tracing/overview.mdx]:
```suggestion
| `header_name` |                                                               | | Enable or disable stdout logging.                                      |
```

File: docs/source/configuration/telemetry/exporters/tracing/overview.mdx [hunk 7]
```diff
@@ -144,9 +144,32 @@ telemetry:
...
            header_name: my-trace-id
+           # The format of the trace when propagating to subgraphs.
+           format: uuid
 ```
 
+#### `request` configuration reference
+
+| Option        | Values                                                        | Default                           | Description                         |
+|---------------|---------------------------------------------------------------|-----------------------------------|-------------------------------------|
+| `header_name` |                                                               | The name of the http header to use for propagation. |                                     |
```
Comment (BrynCooke) [inline docs/source/configuration/telemetry/exporters/tracing/overview.mdx:157]:
```suggestion
| `header_name` |                                                               |  |  The name of the http header to use for propagation.                                   |
```

**Review by garypen — APPROVED** [review]

**Comment by goto-bus-stop** [comment]
This adds a new required config property, so previous configurations are now invalid. Is that intentional?

**Comment by goto-bus-stop** [comment]
The test added in https://github.com/apollographql/router/commit/7b8d801b5f69334454f970d9ff77663545c42a02#diff-6068839267e9a05b915684b22f2a0fc86168216efc92da3113cac6f0d8de8890 fails for example.

**Comment by goto-bus-stop** [comment]
Making it optional and fixing the related test failure on `dev`: https://github.com/apollographql/router/pull/5824

**Pull request merged**